src/dune_pkg/archive_driver.ml

```diff
@@ -0,0 +1,106 @@
+open Stdune
+module Process = Dune_engine.Process
+open Fiber.O
+
+module Command = struct
+  type t =
+    { bin : Path.t
+    ; args : archive:Path.t -> target_in_temp:Path.t -> string list
+    }
+end
+
+type t =
+  { command : Command.t Lazy.t
+  ; suffixes : string list
+  }
+
+let which bin_name = Bin.which ~path:(Env_path.path Env.initial) bin_name
+
+let tar_args ~archive ~target_in_temp =
+  [ "xf"; Path.to_string archive; "-C"; Path.to_string target_in_temp ]
+;;
+
+let zip_args ~archive ~target_in_temp =
+  [ Path.to_string archive; "-d"; Path.to_string target_in_temp ]
+;;
+
+let tar =
+  let command =
+    lazy
+      (match
+         (* Test for tar before bsdtar as tar is more likely to be installed
+            and both work equally well for tarballs. *)
+         List.find_map [ "tar"; "bsdtar" ] ~f:which
+       with
+       | Some bin -> { Command.bin; args = tar_args }
+       | None -> Dune_engine.Utils.program_not_found "tar" ~loc:None)
+  in
+  { command; suffixes = [ ".tar"; ".tar.gz"; ".tgz"; ".tar.bz2"; ".tbz" ] }
+;;
+
+let zip =
+  let command =
+    lazy
+      (match which "unzip" with
+       | Some bin -> { Command.bin; args = zip_args }
+       | None ->
+         (* Fall back to using tar to extract zip archives, which is possible in some cases. *)
+         (match
+            (* Test for bsdtar before tar, as if bsdtar is installed then it's
+               likely that the tar binary is GNU tar which can't extract zip
+               archives, whereas bsdtar can. If bsdtar is absent, try using the
+               tar command anyway, as on BSD, MacOS, and Windows systems, the tar
+               command can extract zip archives. *)
+            List.find_map [ "bsdtar"; "tar" ] ~f:which
+          with
+          | Some bin -> { Command.bin; args = tar_args }
+          | None ->
+            (* Still reference unzip in the error message, as installing it
+               is the simplest way to fix the problem. *)
+            Dune_engine.Utils.program_not_found "unzip" ~loc:None))
+  in
+  { command; suffixes = [ ".zip" ] }
+;;
+
+let all = [ tar; zip ]
+let output_limit = 1_000_000
+let check_suffix t filename = List.exists t.suffixes ~f:(Filename.check_suffix filename)
+let choose_for_filename filename = List.find all ~f:(fun t -> check_suffix t filename)
+
+let extract t ~archive ~target =
+  let* () = Fiber.return () in
```

Comment:
:question:  _Side question_: do you know why we have to do this? I have seen this in various part of the code, I'm curious about the why...

Reply:
I always assumed this with `let+` is so that a `Monad.return` can be avoided in the end.

Reply:
This stops the parts of the fiber from being executed eagerly. If we didn't do this then all those `Lazy.force`s and other stuff would be run before we really run the fiber. The best way to think about it is "from now on, everything that happens happens only when this fiber is run".